### ISSUE TITLE
Request - Add parameter "--delay" to allow web-ext to wait until a WebExtension build has completed the building process

### ISSUE THREAD

**Issue by tommai78101** [mozilla/web-ext#1573]
### Is this a feature request or a bug?

This is a feature.

### What is the current behavior?

The current behavior is, if `web-ext` is launched normally, it will immediately reload any files that were changed. If an extension is in the process of compiling/building, `web-ext` will constantly reload and reload and reload, and may not fully reload completely due to many files being changed in a short period of time.

The other behavior is, if `web-exit` is launched with the parameter, `--no-reload`, then the user needs to manually reload the extension, whenever the compilation process is completed. However, it is inconvenient to have to manually reload the finished extension every time, and this cannot be automated.

### What is the expected or desired behavior?

The desired behavior is by setting a new parameter, `--delay` and specify a given amount of milliseconds (for accurate timing, which some extensions and production builds may benefit from). This delay will allow `web-ext` to wait for a certain period of time duration that was specified, before executing a `reload` command.

Allowing such process would benefit automated production builds, while allowing the flexibility to wait until the compilation process is completed, and can be independently tweaked to the user's content.

### Example command syntax:

`web-ext reload --delay 5000`

`web-ext` will reload after 5 seconds has elapsed.

**Comment by freaktechnik** [comment]
Optimally you'd launch web-ext after your builds were successful or integrate it directly into the chain, like https://github.com/TheAifam5/webext-webpack-plugin

Further, my understanding is that reloads are debounced to a certain degree.

**Comment by Rob--W** [comment]
Currently we've hard-coded the delay to 1 second.
Supporting an option, such as `--reload-delay` sounds good.

We just need to add a parameter and forward the parameter to the following location:
https://github.com/mozilla/web-ext/blob/588550008c1d0381736f435956095255200707d5/src/watcher.js#L39
... and add unit tests.

**Comment by kumar303** [comment]
This is the same problem as https://github.com/mozilla/web-ext/issues/1137 and I don't think adding a delay is a good solution. It will work on some machines but not others. It will make QA / test engineers cry.

Maybe it's time that `web-ext` ships with an official webpack plugin? Or perhaps we should officially endorse one?

**Comment by Rob--W** [comment]
The current 1 second limit is also arbitrary. A configurable debounce delay is a *simple* and general way to reduce the number of unwanted autoreloads. A webpack plugin would be surgically precise, but it is also more complicated to set up (and specific to webpack).
Both options can co-exist.

If we are more interested in solving #1137 in a generic way, then I would consider adding an option to only watch a specific file, with the following flow:

1. `web-ext run` starts, starts watching.
2. Build process starts, clears build dir.  *We should ignore when the file does not exist, to avoid reloading when the extension dir is empty.*
3. Build process finishes, touches clobber file. *`web-ext` detects the file change, and reloads the extension*

**Comment by caitmuenster** [comment]
If this is your first contribution, please refer to https://github.com/mozilla/web-ext/blob/master/CONTRIBUTING.md for how to get started. 

Mentor: @Rob--W

**Comment by suhailsinghbains** [comment]
Hi, I would like to work on this bug.
Thanks

**Comment by Rob--W** [comment]
Hi @suhailsinghbains, welcome!

Just to recap, the feature request here is to add a flag to allow users of `web-ext` to force `web-ext` to wait for at least the given amount of time before reloading Firefox in response to file changes.

The current behavior is as follows:

1. Upon file change, Firefox is immediately reloaded (because of [`executeImmediately = true` at `src/watcher.js`](https://github.com/mozilla/web-ext/blob/88cbd12b135183dab5229775bb784654b23e47d9/src/watcher.js#L38-L39)).
2. Upon further file changes, nothing happens until at least 1 second (1000 milliseconds) have passed.

The changes needed to implement the requested feature is:

1. Add a new flag called `reload-delay` after [the `no-reload` option definition in `src/program.js`](https://github.com/mozilla/web-ext/blob/88cbd12b135183dab5229775bb784654b23e47d9/src/program.js#L471-L475).
2. Support the new `reloadDelay` option at [`src/cmd/run.js`](https://github.com/mozilla/web-ext/blob/88cbd12b135183dab5229775bb784654b23e47d9/src/cmd/run.js) (see the existing `noReload` for an example of how you can use the parameter).
3. In [`src/extension-runners/index.js`](https://github.com/mozilla/web-ext/blob/358a63ad6bf0f1ea8964daa5848ac25c3695ebe4/src/extension-runners/index.js#L258), define the `reloadDelay` parameter in the `ReloadStrategyParams` type, and forward the value to the `createWatcher` call. Also add the `reloadDelay` option to `WatcherCreatorParams` and forward it in `defaultWatcherCreator` to `onSourceChange`.
4. In [`src/watcher.js`](https://github.com/mozilla/web-ext/blob/358a63ad6bf0f1ea8964daa5848ac25c3695ebe4/src/watcher.js#L17-L39), define the `reloadDelay` option in `OnSourceChangeParams`, and let `onSourceChange` check if the parameter is set. If it is set, let `onChange` be `debounce(onChange, reloadDelay)`. Otherwise fall back to the current behavior (i.e. `debounce(onChange, 1000, reloadImmediately)`).
5. Add unit tests to verify that the implementation works as expected. See `test/unit/test-cmd/test.run.js` for examples. Put the new tests after [the tests of `--no-reload`](https://github.com/mozilla/web-ext/blob/44114ee23af82abc1ec76f9770a0ccb452c6e29b/tests/unit/test-cmd/test.run.js#L189).

**Comment by suhailsinghbains** [comment]
Thanks a lot :)

**Comment by icl7126** [comment]
@Rob--W what about your suggestion with watching only single file? That sounds like a much better solution as it is reactive, reliable and easy to use (modifying single file once the build is done is piece of cake).

**Comment by Rob--W** [comment]
@icl7126 Could you open a new issue and describe your use case? For context, you could reference https://github.com/mozilla/web-ext/issues/1573#issuecomment-482526896 . These two features can co-exist, and I'd like to hear from other team members whether they also think whether this is a good idea.

**Comment by kumar303** [comment]
Is it really helpful to add a `--reload-delay` param to `web-ext`? It will just replace *a lot* of flakiness with *some* flakiness. The delay won't be correct for everyone's machine.

Could we somehow detect a parallel build process and warn, like, "reloading will be more reliable if you hook into your current build process"? No solution immediately comes to mind for detecting that but maybe there is a way. (Maybe just look for any `webpack.*` file in the current directory?)

At the very least, the docs for this option should say it's a bad idea to rely on it and they should link to existing webpack plugins that can be used to solve the problem correctly. Also, when calling this option [programatically](https://github.com/mozilla/web-ext/#using-web-ext-in-nodejs-code) it would be nice to emit a warning so that people don't go and write scripts with delay hacks unless they have no other option.

**Comment by kumar303** [comment]
Actually, `web-ext` could emit a warning any time `--reload-delay` was set. The warning could offer links to webpack plugins. Maybe that would be enough to discourage using it.
